rgw: migrate from boost::variant to std::variant

tchaikov opened the pull request:
replacing boost::variant with std::variant throughout the rgw codebase. This change is part of our ongoing effort to reduce third-party dependencies by leveraging C++ standard library alternatives where possible.

Benefits include:
- Improved code readability and maintainability
- Reduced external dependency surface
- More consistent API usage with other components





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

mattbenjamin commented:
I think replacing boost::variant with std::variant doesn't have a large benefits being claimed.  I don't mind doing it, but I somewhat object to the propaganda

mattbenjamin commented on src/rgw/rgw_file.cc:
```diff
@@ -1825,6 +1824,33 @@ namespace rgw {
     }
   }
 
+  bool RGWListBucketsRequest::eof() {
```
this changes removes the inline property from the eof() method, and I don't want to do that

tchaikov replied:
i added the `inline` specifier.

mattbenjamin replied:
perfect

mattbenjamin commented on src/rgw/rgw_file.cc:
```diff
@@ -1825,6 +1824,33 @@ namespace rgw {
+    return !is_truncated && !rcb_eof;
+  }
+
+  bool RGWReaddirRequest::eof() {
```
same here

mattbenjamin reviewed (changes requested):
please restore the inlining of RGWListBucketsRequest::eof() and RGWReaddirRequest::eof()

mattbenjamin commented:
to clarify, I support the change to std::variant, it's always generally preferable to use the std forms when comparable

tchaikov commented:
> I think replacing boost::variant with std::variant doesn't have a large benefits being claimed. I don't mind doing it, but I somewhat object to the propaganda

i dropped the "benefits" section from the commit message. hope it's less exaggerating now.

tchaikov commented:
> please restore the inlining of RGWListBucketsRequest::eof() and RGWReaddirRequest::eof()

done.

tchaikov commented:
v2:

- dropped the "benefits" section in the commit message
- added `inline` specifier to the previously inlined functions
- added a dedicated commit to move the `eof()` member functions to .cc file. the commit is accompanied with the commit message explaining why we should extract them into .cc file.

tchaikov commented:
@mattbenjamin hi Matt, thank you for your review. could you please take another look?

tchaikov commented:
jenkins test make check

tchaikov commented:
```
apt-get install -y clang-19 lldb-19 lld-19 clangd-19
Reading package lists...
Building dependency tree...
Reading state information...
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 python3-lldb-13 : Conflicts: python3-lldb-x.y
 python3-lldb-19 : Conflicts: python3-lldb-x.y
E: Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
Build step 'Execute shell' marked build as failure
```

see https://jenkins.ceph.com/job/ceph-pull-requests/158067/consoleFull#724108817fa534114-2506-4c16-be9c-09c932a5aa49

tchaikov commented:
@mattbenjamin hi Matt, could you please take a look?

awojno-bloomberg commented on src/rgw/rgw_file_int.h:
```diff
@@ -615,8 +613,7 @@ namespace rgw {
 		    uint8_t obj_type) {
-      using std::get;
-      directory* d = get<directory>(&variant_type);
+      directory* d = get_if<directory>(&variant_type);
```
Can we add `std::` to the `get_if` calls for consistency?

tchaikov commented:
changelog:

- add `std::` qualifier to `get_if()` calls to be more consistent.

mattbenjamin reviewed (approved):
lgtm

cbodley commented:
passed qa in https://pulpito.ceph.com/cbodley-2025-06-25_18:43:57-rgw-wip-cbodley-testing-distro-default-smithi/
